File: C++/1578_MinimumTimeToMakeRopeColorful.cpp
// Name: Naman Goyal
// Date: 4/10/2022

class Solution {
  public:
    int minCost(string colors, vector < int > & neededTime) {
      int n = colors.size();
      int ans = 0;
      // Iterating over the colors array and checking the equal colors values 
      for (int i = 0; i < n; i++) {
        int j = i;
        int sum = 0, ma = 0;
        while (j < n && colors[j] == colors[i]) {
          sum += neededTime[j];
          ma = max(ma, neededTime[j]);
          j++;
        }
        j--;
        // Finding the maximum value and subtracting it from the sum if no of same color values >= 2, taking only one value
        if (j - i + 1 >= 2) ans += sum - ma;
        i = j;
      }
      return ans;
    }
};

File: PYTHON/1578_MinimumTimeToMakeRopeColorful.py
# NAME : Shubham Arora
# DATE : 23/Oct/2022
class Solution:
    def minCost(self, colors: str, neededTime: List[int]) -> int:
        t = neededTime[0]
        s = 0
        for i in range(1 , len(colors)):
            # finding the maximum of the needed time for both the elements in case they are equal
            if colors[i] == colors[i - 1]:
                t=  max(t , neededTime[i])
            else:
                # adding it to the sum if the elements are not equal
                s += t
                # reinitializing t after every iteration
                t = neededTime[i]
        # adding the time for last element in the sum
        s += t
        # returning the final answer
        return sum(neededTime) - s
